### PR TITLE
[SDK] Add retry for cloud serving credential and workspace connection calls

### PR DESCRIPTION
**Pull request by brynn-code**
# Description

- Use managed identity with retry for cloud serving.
- Add retry when getting workspace connections in local.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_serving/extension/azureml_extension.py [hunk 1]
```diff
@@ -175,5 +182,21 @@ def _initialize_connection_provider(self):
...
+            self.connection_provider = AML_CONNECTION_PROVIDER_TEMPLATE.format(
+                self.subscription_id, self.resource_group, self.workspace_name
+            )  # noqa: E501
+
+
+def _get_managed_identity_credential_with_retry(**kwargs):
+    from azure.identity import ManagedIdentityCredential
+
+    class ManagedIdentityCredentialWithRetry(ManagedIdentityCredential):
+        @retry(Exception)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_serving/extension/azureml_extension.py:198]: what's the default retry number and machanism?

Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/_serving/extension/azureml_extension.py:198]: default retry count is 4, and retry intervals are 3, 6, 12. (delay(3) * backoff(2)^retry_time)

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 2]
```diff
@@ -36,6 +36,9 @@ def __init__(self, **kwargs):
         self._runs = RunOperations()
         self._connection_provider = None
         self._config = kwargs.get("config", None) or {}
+        # The credential is used as an optional to override
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_pf_client.py:39]: an optional?

Comment (brynn-code) [reply src/promptflow/promptflow/_sdk/_pf_client.py:39]: good catch, my fingers have their own muscle memory

**Review by wangchao1230 — APPROVED** [review]

**Review by zhengfeiwang — APPROVED** [review]

**Pull request merged**